### PR TITLE
Review

### PR DESCRIPTION
**Pull request by qweliant**
This adds a new way to solicit reviews. 

Try this as a logged in user :

visit a draft pub. copy the review share link from the sharing tab(hold onto that link)
paste the link your browser and submit a review
you should be navigated to the reviews dashboard

visit a released pub. click the review link from the header
paste the link your browser and submit a review
you should be navigated to the reviews dashboard

Try this as a logged out user:
use the link from earlier
paste the link your browser and submit a review
you should be navigated to the signup page

### CODE REVIEW

File: client/containers/Pub/PubDocument/Review/ReviewEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	setReviewDoc(review);
+
+	const updatingReviewDoc = (doc: DocJson) => {
+		setReview(doc);
+		setReviewDoc(review);
+	};
+
+	useEffect(() => {
+		setReviewDoc(review);
+	}, [review, setReviewDoc]);
```
[the remark this answers is not in the document]
Comment (howebenhowe) [reply client/containers/Pub/PubDocument/Review/ReviewEditor.tsx:40]: And now that I think about it, if this is in the parent `<Review>` component, it won't even need to `useState`, since that state management is handled in `useLocalStorage`

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+const ReviewModal = (props: Props) => {
+	const { isLoading, pubData, communityData, updatePubData, reviewDoc, setIsLoading } = props;
+	const {
+		scopeData: { activePermissions },
+		loginData: { fullName },
+	} = usePageContext();
+	const [reviewTitle, setReviewTitle] = useState('Untilted Review');
+	const [reviewerName, setReviewerName] = useState(fullName || 'anonymous');
+	const [createError, setCreateError] = useState(undefined);
+	const redirectCondition = activePermissions.canEdit || fullName;
```
Comment (howebenhowe) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:29]: maybe more explicit with
```
const isLoggedIn = !!fullName;
const redirectCondition = activePermissions.canEdit || isLoggedIn;
```

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+						reviewerName={reviewerName}
+						setReviewerName={setReviewerName}
+						canEdit={activePermissions.canEdit}
+					/>
+				)}
+			</DialogLauncher>
+
+			{createError && (
+				<NonIdealState
+					title="Something something errors"
```
Comment (howebenhowe) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:102]: Luvit. Could perhaps be more specific.

File: client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	reviewTitle: string;
+	reviewerName: any;
+	setReviewerName: any;
+	canEdit: boolean;
+};
+
+const defaultProps = {};
+
+type Props = OwnProps & typeof defaultProps;
+
```
Comment (howebenhowe) [inline client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx:23]: What does `typeof defaultProps` do here, when defaultProps is {}? Not a rhetorical question, I really don't know what this does.

File: server/review/api.ts [hunk 5]
```diff
@@ -15,9 +15,11 @@ const getRequestIds = (req) => {
 
 app.post('/api/reviews', (req, res) => {
 	const requestIds = getRequestIds(req);
+	const { permissions } = req.body;
 	getPermissions(requestIds)
-		.then((permissions) => {
-			if (!permissions.create) {
+		// how should permissions exist on a api call that can be used by anyone?
```
Comment (howebenhowe) [inline server/review/api.ts:20]: Dunno!
I defer on resolving this to press on with the rest of the review, but don't want it to slip through.

**Review by howebenhowe — CHANGES_REQUESTED** [review]
Several suggestions throughout, but the code looks great! I'll get to testing per your PR's instruction next.

**Review by howebenhowe — CHANGES_REQUESTED** [review]
Several suggestions throughout, but the code looks great! I'll get to testing per your PR's instruction next.

File: client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx (hunk 4 above)
Comment (qweliant) [reply client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx:23]: me niether its a copy paste thing 🤷🏿‍♂️

File: client/containers/DashboardReviews/DashboardReviews.tsx [hunk 6]
```diff
@@ -16,6 +16,13 @@ const DashboardReviews = (props: Props) => {
 	const { pubsWithReviews } = props;
 	const { scopeData } = usePageContext();
 	const { activeCollection, activeTargetType } = scopeData.elements;
+	const parseReviewers = (reviewers: Reviewer[] | undefined) => {
```
[the remark this answers is not in the document]
Comment (howebenhowe) [reply client/containers/DashboardReviews/DashboardReviews.tsx:19]: Rereading my comment above, I misread 'in-competent' three times before I could get it right.

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	setIsLoading: any;
+};
+
+const ReviewModal = (props: Props) => {
+	const { isLoading, pubData, communityData, updatePubData, reviewDoc, setIsLoading } = props;
+	const {
+		scopeData: { activePermissions },
+		loginData: { fullName },
+	} = usePageContext();
+	const [reviewTitle, setReviewTitle] = useState('Untilted Review');
```
Comment (howebenhowe) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:26]: Untilted, still made me smile, but I suspect we want 'untitled'

**Review by howebenhowe — CHANGES_REQUESTED** [review]
uno mas

**Comment by howebenhowe** [comment]
Errant behavior from testing script above: the local storage review copy was not erased after submitting the pub.
_steps_
1. Began a review as logged-in user.
2. paused mid-review to refresh and check the local storage of in-progress review. Works great
3. finish and submit review
4. success
5. log out
6. paste review link
7. on arriving to the review page, review space is already populated with copy from above logged-in review

File: client/components/PubShareDialog/PubShareDialog.tsx [hunk 8]
```diff
@@ -68,14 +71,32 @@ const AccessHashOptions = (props: AccessHashOptionsProps) => {
...
+		pubUrl(communityData, pubData, {
+			accessHash: hash,
+			historyKey: historyData.currentKey,
+			isReview: true,
+		});
+
+	const renderRow = (label, hash, urlFunction) => {
+		const url = urlFunction(hash);
+		return renderCopyLabelComponent(label, url);
+	};
```
Comment (3mcd) [inline client/components/PubShareDialog/PubShareDialog.tsx:89]: We may be able to remove the `*UrlFunction` functions and the `renderRow` function by sticking to plain old variables. IMO it makes the code a bit easier to follow! ie.

```ts
const isDraft = !isRelease
const createAccessUrl = (accessHash, options) => pubUrl(communityData, pubData, { accessHash, ...options })
const reviewAccessUrl = getAccessUrl(viewHash, { historyKey: historyData.currentKey, isReview: true })
// ...
{viewHash && renderCopyLabelComponent('View', createAccessUrl(viewHash, { isDraft }))}
{editHash && renderCopyLabelComponent('Edit', createAccessUrl(editHash, { isDraft }))}
{renderCopyLabelComponent('Review', reviewAccessUrl)}
```

File: client/containers/DashboardReviews/DashboardReviews.tsx [hunk 9]
```diff
@@ -66,6 +76,11 @@ const DashboardReviews = (props: Props) => {
 													<div className="number">
 														R{review.number}: {review.title}
 													</div>
+													{hasReviewers && (
+														<div className="note">
+															by {parseReviewers(reviewers)}&nbsp;
```
Comment (3mcd) [inline client/containers/DashboardReviews/DashboardReviews.tsx:81]: IMO this function should be called `renderReviewerByline` or something similar since no actual parsing, ie transforming data into a higher level representation, occurs within

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+		return apiFetch
+			.post('/api/reviewer', {
+				reviewId: review.id,
+				name: reviewerName,
+				permissions: activePermissions,
+			})
+			.catch((err) => {
+				setIsLoading(false);
+				setCreateError(err);
+			});
```
Comment (3mcd) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:42]: Nice job adding error handling/states to this feature!

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+			})
+			.then((review) => {
+				saveReviewerName(review);
+				updatePubData((currentPubData) => {
+					return {
+						reviews: [...currentPubData.reviews, review],
+					};
+				});
+				setIsLoading(false);
+				window.location.href = redirectUrl(review);
```
Comment (3mcd) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:62]: Do we ever let anonymous users know that they'll be redirected to the signup screen when their review is submitted? It may be a little bewildering to people why they ended up there without any notice.

File: client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+					Create Review
+				</Button>
+			</React.Fragment>
+		);
+	};
+	const helperTextToRender = (
+		<p>
+			You are about to submit a review for <b>{pubData.title}</b>.
+		</p>
+	);
```
Comment (3mcd) [inline client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx:47]: Should we just move this into the JSX block below since it's fully declarative/markup?

File: client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+					<div className="title-input">
+						<p>Add a title to your review?</p>
+						<InputGroup
+							defaultValue={reviewTitle}
+							onKeyDown={(evt) => {
+								if (evt.key === 'Enter') {
+									evt.currentTarget.blur();
+								}
+							}}
+							onBlur={(evt) => setReviewTitle(evt.target.value.trim())}
```
Comment (3mcd) [inline client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx:68]: These event handlers (and below) may be good candidates to move out of this block to keep the visual noise down!

File: server/routes/pubDocument.tsx [hunk 14]
```diff
@@ -235,3 +251,42 @@ app.get(
 		}
 	},
 );
+
+app.get(['/pub/:pubSlug/review/:historyKey/'], async (req, res, next) => {
+	if (!hostIsValid(req, 'community')) {
+		return next();
+	}
+	try {
```
Comment (3mcd) [inline server/routes/pubDocument.tsx:259]: We have a `wrap` helper that should let us get rid of the try/catch here. But I see we haven't used it in a few other routes. Did you look into this already?

File: types/user.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
+export type MinUser = {
```
[the remark this answers is not in the document]
Comment (3mcd) [reply types/user.ts:1]: ++ for `MinimalUser`

**Review by 3mcd — CHANGES_REQUESTED** [review]
Awesome work @qweliant!! I still need to take a look at the feature in-browser, but I don't expect I'll find anything major.

Most of my suggestions here are cosmetic, but I do think they'll help with readability/maintainability.

File: server/reviewer/queries.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
+import { Reviewer } from 'server/models';
+
+export const createReviewer = (inputValues) => {
+	return Reviewer.create({
```
Comment (idreyn) [inline server/reviewer/queries.ts:4]: I think you can write:

```ts
return Reviewer.create({ ... }, { returning: true });
```

 to get Sequelize to return the created model so you don't have to re-query it (it's a little odd that they don't do that by default.)

File: server/reviewer/queries.ts [hunk 17]
```diff
@@ -0,0 +1,12 @@
+import { Reviewer } from 'server/models';
+
+export const createReviewer = (inputValues) => {
```
Comment (idreyn) [inline server/reviewer/queries.ts:3]: In `createReview` I noticed that you destructured `inputValues` near the beginning of the function to make it clear what needs to be provided there...if I had to guess that was because you had to carefully read the function body in order to tell what values were required!

Here and now you have the power to forestall a similar fate for your colleagues by either destructuring in the function signature or (better) declaring a type for `inputValues` :)

File: server/reviewer/model.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
+export default (sequelize, dataTypes) => {
+	return sequelize.define(
+		'Reviewer',
+		{
+			id: sequelize.idType,
+			name: { type: dataTypes.TEXT },
```
Comment (idreyn) [inline server/reviewer/model.ts:6]: $0.02: I'd suggest `fullName` for symmetry with the `User` model.

File: server/routes/pubDocument.tsx [hunk 19]
```diff
@@ -81,12 +85,24 @@ const getEnrichedPubData = async ({
...
 		getDraft: true,
 		getDiscussions: true,
+		isReview,
 	});
 
 	if (!pubData) {
 		throw new ForbiddenError();
 	}
 
+	if (!pubData.reviewHash) {
```
Comment (idreyn) [inline server/routes/pubDocument.tsx:95]: Hmm, I dunno about this. It's more typical to write a migration to backfill these instead. (I appreciate the attention to detail, but if we took this approach everywhere our code would be thick with these ad-hoc backfills by now!)

File: server/routes/pubDocument.tsx [hunk 20]
```diff
@@ -235,3 +251,42 @@ app.get(
 		}
 	},
 );
+
+app.get(['/pub/:pubSlug/review/:historyKey/'], async (req, res, next) => {
```
Comment (idreyn) [inline server/routes/pubDocument.tsx:255]: Perhaps this be merged into the Pub draft handler above in the following way:

1. Adding `'/pub/:pubSlug/review/:historyKey/' to the array of routes it handles
2. Checking `req.path` to see whether we're handing the Pub in review mode

I encourage you to give this some thought and then decide for yourself whether this is a good idea. It could also be premature to merge these two handlers, since their behavior may diverge sharply in the future! In this case, since they are very nearly identical and there is a lot of tricky logic that has to be duplicated between them, I would probably merge them — or at least factor some of the shared `historyKey` munging into a helper fn.

File: tools/migrations/2022_07-19_noNullUserIdReviewNew.js [hunk 21]
```diff
@@ -0,0 +1,13 @@
+export const up = async ({ Sequelize, sequelize }) => {
```
Comment (idreyn) [inline tools/migrations/2022_07-19_noNullUserIdReviewNew.js:1]: Sorry, sorry, but I have to tell you that this file is named inconsistently with all the rest of them, using a dash instead of an underscore!

File: package.json [hunk 22]
```diff
@@ -105,6 +105,7 @@
 		"keyboardjs": "^2.5.1",
 		"knex": "^0.21.17",
 		"linkifyjs": "^3.0.0-beta.3",
+		"lodash.defer": "^4.1.0",
```
Comment (idreyn) [inline package.json:108]: ⚠️ Do you end up using this anywhere? If not, let's remove it.

File: client/containers/Pub/PubHeader/DraftReleaseButtons.tsx [hunk 23]
```diff
@@ -64,43 +64,60 @@ const DraftReleaseButtons = (props: DraftReleaseButtonsProps) => {
...
+				<ResponsiveHeaderButton
+					icon="draw"
+					tagName="a"
+					href={pubUrl(communityData, pubData, {
+						accessHash: reviewHash,
+						historyKey: historyData.currentKey,
+						isReview: true,
+					})}
+					outerLabel={{
+						top: 'Create a review of this pub',
```
Comment (idreyn) [inline client/containers/Pub/PubHeader/DraftReleaseButtons.tsx:76]: Don't forget to capitalize Pub in user-facing copy.

File: client/containers/Pub/PubDocument/Review/ReviewEditor.tsx [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+
+	return (
+		<div className="review-editor">
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/ReviewEditor.tsx:43]: is this wrapper div doing anything?

File: client/containers/Pub/PubDocument/Review/reviewEditor.scss [hunk 25]
```diff
@@ -0,0 +1,6 @@
+.review-editor {
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/reviewEditor.scss:1]: Re: the wrapper div from earlier — it's our convention to call the outermost class in a React component `.my-thing-component` instead of just `.my-thing`.

File: client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+
+require('./reviewerDialog.scss');
+
+type Props = {
+	isOpen: boolean;
+	pubData: {
+		title: string;
+	};
+	onClose: (...args: any[]) => any;
+	createReviewDoc: () => any;
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx:12]: Conventionally, event handlers passed into a component start with `on`. So this would be `onCreateReviewDoc`.

File: client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx [hunk 27]
```diff
@@ -0,0 +1,95 @@
+import React from 'react';
+import { Button, Classes, Dialog, InputGroup } from '@blueprintjs/core';
+
+require('./reviewerDialog.scss');
+
+type Props = {
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/ReviewerDialog.tsx:6]: Please do a pass on these to provide props that are more specific than `any` in each case.

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+				)}
+			</DialogLauncher>
+
+			{createError && (
+				<NonIdealState
+					title="There was an error submitting your review, login or signup to contact a community member"
+					// @ts-expect-error ts-migrate(2322) FIXME: Type '{ title: string; visual: string; action: Ele... Remove this comment to see the full error message
+					visual="error"
+					action={
+						<a href="/login?redirect=/community/create" className="bp3-button">
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:106]: This would then send a user onward to create a Community, which I don't think is the intention here!

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+						reviewerName={reviewerName}
+						setReviewerName={setReviewerName}
+						isUser={isUser}
+					/>
+				)}
+			</DialogLauncher>
+
+			{createError && (
+				<NonIdealState
+					title="There was an error submitting your review, login or signup to contact a community member"
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:102]: Assuming they login or signup — what would be the next step for contacting a Community member? I don't know if this suggestion will quite get them where they need to go.

File: client/containers/Pub/PubDocument/Review/ReviewModal.tsx [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+import { Button, NonIdealState } from '@blueprintjs/core';
+
+import { apiFetch } from 'client/utils/apiFetch';
+import { DialogLauncher } from 'components';
+import { DocJson } from 'types';
+import { usePageContext } from 'utils/hooks';
+
+import ReviewerDialog from './ReviewerDialog';
+
+type Props = {
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/Review/ReviewModal.tsx:11]: Let's please grab the appropriate types from the `types` directory — `Pub` and `Community` are really easy to import, and you might also be interested in the `Callback` type.

```ts
// For a callback with no arguments
type Callback = () => unknown;

// For a callback with an argument of type T
type Callback<T> = (t: T) => unknown;
```

In your case you could e.g. write `updatePubData: Callback<Partial<Pub>>`

File: client/containers/Pub/PubDocument/PubHistoricalNotice.tsx [hunk 31]
```diff
@@ -25,12 +25,12 @@ type Props = {
 const PubHistoricalNotice = (props: Props) => {
 	const { pubData, historyData } = props;
 	const { communityData } = usePageContext();
-	const { releases, releaseNumber, isRelease } = pubData;
+	const { releases, releaseNumber, isRelease, isReview } = pubData;
```
Comment (idreyn) [inline client/containers/Pub/PubDocument/PubHistoricalNotice.tsx:28]: From `PubDocument` it looks like we skip showing this component when we have a review on our hands, so maybe we can skip these changes?

File: client/components/PubShareDialog/PubShareDialog.tsx [hunk 32]
```diff
@@ -68,14 +71,32 @@ const AccessHashOptions = (props: AccessHashOptionsProps) => {
...
+
 	return (
 		<div className="access-hash-options">
 			<p>
-				You can grant visitors permission to view or edit the draft of this pub by sharing a
-				URL.
+				You can grant visitors permission to view, edit, or review the draft of this pub by
```
Comment (idreyn) [inline client/components/PubShareDialog/PubShareDialog.tsx:94]: Not your doing but would you mind capitalizing Pub while you're in here? :)

File: client/components/PubShareDialog/PubShareDialog.tsx [hunk 33]
```diff
@@ -12,15 +12,18 @@ import {
...
 import { useMembersState } from 'client/utils/members/useMembers';
 import { pubUrl } from 'utils/canonicalUrls';
+import { usePubContext } from 'containers/Pub/pubHooks';
 
 require('./pubShareDialog.scss');
 
 type PubShareDialogProps = {
 	isOpen: boolean;
 	onClose: (...args: any[]) => any;
 	pubData: {
```
Comment (idreyn) [inline client/components/PubShareDialog/PubShareDialog.tsx:22]: Might be good to just bring the `Pub` type in here (these prop definitions were likely written before it existed.)

**Review by idreyn — CHANGES_REQUESTED** [review]
There is a lot of really good work in here, I love how it builds on the existing Reviews feature, and I'm excited to see this launched! I know I just added a _lot_ of comments, but I'm mostly concerned about the ones marked 🚨 where there's a security hole. Happy to talk those through in DMs if you want. 

Other than that, I'd appreciate a look at the rest, but your velocity on this has been great and I don't want to slow you down all that much with my usual grab-bag of pedantry. Shipping beats perfection and there will be time to clean up parts of this later.

File: server/pub/permissions.ts [hunk 34]
```diff
@@ -23,6 +23,7 @@ const managerUpdatableFields = [
 	'nodeLabels',
 	'pubEdgeListingDefaultsToCarousel',
 	'pubEdgeDescriptionVisible',
+	'reviewHash',
```
[the remark this answers is not in the document]
Comment (qweliant) [reply server/pub/permissions.ts:26]: let me think for a sec...i was trying to set a reviewHash on the client if it doesnt exist. but then i changed that to be on the server, so a review hash is ensured for a pub. so yeah this can go bc these are fields that be can be updated by users

File: tools/migrations/2022_07-19_noNullUserIdReviewNew.js (hunk 21 above)
Comment (qweliant) [reply tools/migrations/2022_07-19_noNullUserIdReviewNew.js:1]: lol good catch

File: client/containers/DashboardReviews/DashboardReviews.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (qweliant) [reply client/containers/DashboardReviews/DashboardReviews.tsx]: lmao, i think it can stay bc otherwise i would have to set reviewers on the server, which isnt a bad idea. but since we pull in reviews and iterate there i think its easier to leave them

File: client/containers/DashboardReviews/DashboardReviews.tsx [hunk 35]
```diff
@@ -50,6 +57,9 @@ const DashboardReviews = (props: Props) => {
 							{pub.reviews
 								.sort((a, b) => Date.parse(b.createdAt) - Date.parse(a.createdAt))
 								.map((review) => {
+									const hasReviewers =
+										review.reviewer && review.reviewer.length > 0;
+									const reviewers = review.reviewer;
```
[the remark this answers is not in the document]
Comment (qweliant) [reply client/containers/DashboardReviews/DashboardReviews.tsx:62]: i want to wait on that. it would be nice to do exactly that but we gotta think about how a user can be represented as an attribution. im thinking as a MinimulUser or User. Instead of attributable user for pub or collection and reviewer. for now im fine with it

File: client/containers/Pub/PubDocument/Review/ReviewEditor.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (qweliant) [reply client/containers/Pub/PubDocument/Review/ReviewEditor.tsx:40]: yeah, but i have to call the localStorage within `ClientOnly`. and i would rather not wrap the entire review component

[the remark this answers is not in the document]
Comment (idreyn) [reply client/containers/Pub/PubDocument/Review/ReviewEditor.tsx:40]: Hmm, that's annoying. Maybe I should add a server-side fallback to `useLocalStorage()` instead of forcing consumers to do that.

File: package.json [hunk 36]
```diff
@@ -34,7 +34,7 @@
 		"workers-prod": "SEQUELIZE_MAX_CONNECTIONS=1 NODE_PATH=./dist/server/client:./dist/server WORKER=true node dist/server/workers/init"
 	},
 	"dependencies": {
-		"@benrbray/prosemirror-math": "0.2.2",
+		"@benrbray/prosemirror-math": "^0.2.2",
```
Comment (idreyn) [inline package.json:37]: Let's revert this change, it's important that we keep this package pegged to a specific version because we're fucking around with its internals so much 😅

**Review by idreyn — APPROVED** [review]
image

**Review by 3mcd — APPROVED** [review]
